### PR TITLE
MINOR: make sure alterAclsPurgatory is closed when controller server …

### PR DESCRIPTION
**Pull request by chia7712**
The `alterAclsPurgatory` can get closed by kafka broker so Kraft should close it as well.

related to #10550

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
@rondagostino FYI

File: core/src/main/scala/kafka/server/ControllerServer.scala [hunk 1]
```diff
@@ -200,6 +200,8 @@ class ControllerServer(
         CoreUtils.swallow(controllerApisHandlerPool.shutdown(), this)
       if (quotaManagers != null)
         CoreUtils.swallow(quotaManagers.shutdown(), this)
+      if (controllerApis != null)
+        CoreUtils.swallow(controllerApis.close(), this)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ControllerServer.scala:203]: Shouldn't this be closed right after the controllerApisHandlerPool?

**Review by ijuma — COMMENTED** [review]
Nice catch. Can we add a test or adjust an existing test to assert that the item has been closed?

File: core/src/main/scala/kafka/server/ControllerServer.scala (hunk 1 above)
Comment (chia7712) [reply core/src/main/scala/kafka/server/ControllerServer.scala:203]: copy that.

**Comment by chia7712** [comment]
@ijuma I had added the unit test. Could you take a look?

File: core/src/main/scala/kafka/server/ControllerApis.scala [hunk 2]
```diff
@@ -74,7 +75,15 @@ class ControllerApis(val requestChannel: RequestChannel,
 
   val authHelper = new AuthHelper(authorizer)
   val requestHelper = new RequestHandlerHelper(requestChannel, quotas, time)
-  private val aclApis = new AclApis(authHelper, authorizer, requestHelper, "controller", config)
+  private[server] val aclApis = new AclApis(authHelper, authorizer, requestHelper, "controller", config)
+  private val _isClosed = new AtomicBoolean(false)
+
+  def isClosed: Boolean = _isClosed.get()
+
+  def close(): Unit = if (_isClosed.compareAndSet(false, true)) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ControllerApis.scala]: Why do we need the compare and set here and the other class?

Comment (chia7712) [reply core/src/main/scala/kafka/server/ControllerApis.scala]: make sure it gets closed only once. I'm fine to remove it if it is overkill :)

File: core/src/main/scala/kafka/server/ControllerApis.scala [hunk 3]
```diff
@@ -74,7 +74,15 @@ class ControllerApis(val requestChannel: RequestChannel,
 
   val authHelper = new AuthHelper(authorizer)
   val requestHelper = new RequestHandlerHelper(requestChannel, quotas, time)
-  private val aclApis = new AclApis(authHelper, authorizer, requestHelper, "controller", config)
+  private[server] val aclApis = new AclApis(authHelper, authorizer, requestHelper, "controller", config)
+  private var _isClosed = false
+
+  def isClosed: Boolean = _isClosed
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ControllerApis.scala]: Can this just return whether aclApis is closed?

File: core/src/main/scala/kafka/server/AclApis.scala [hunk 4]
```diff
@@ -48,10 +48,11 @@ class AclApis(authHelper: AuthHelper,
   this.logIdent = "[AclApis-%s-%s] ".format(name, config.nodeId)
   private val alterAclsPurgatory =
       new DelayedFuturePurgatory(purgatoryName = "AlterAcls", brokerId = config.nodeId)
+  private var _isClosed = false
 
-  def close(): Unit = {
-    alterAclsPurgatory.shutdown()
-  }
+  def isClosed: Boolean = _isClosed
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/AclApis.scala]: Can this just return whether `alterAclsPurgatory` is closed?

**Comment by chia7712** [comment]
`ConsumerBounceTest.testCloseDuringRebalance` is flaky on trunk so it is unrelated to this PR.

**Review by ijuma — APPROVED** [review]
LGTM, thanks!

**Comment by ijuma** [comment]
We should cherry-pick this to 3.0 please.

**Comment by chia7712** [comment]
> We should cherry-pick this to 3.0 please.

will copy that

**Comment by chia7712** [comment]
push to trunk and 3.0

**Pull request merged**